packages/api/src/router/widgets/dns-hole.ts

```diff
@@ -1,44 +1,68 @@
-import { TRPCError } from "@trpc/server";
+import { observable } from "@trpc/server/observable";
 
+import type { Modify } from "@homarr/common/types";
+import type { Integration } from "@homarr/db/schema/sqlite";
+import type { IntegrationKindByCategory, WidgetKind } from "@homarr/definitions";
 import { getIntegrationKindsByCategory } from "@homarr/definitions";
 import { integrationCreator } from "@homarr/integrations";
 import type { DnsHoleSummary } from "@homarr/integrations/types";
-import { logger } from "@homarr/log";
-import { createCacheChannel } from "@homarr/redis";
+import { createItemAndIntegrationChannel } from "@homarr/redis";
+import { z } from "@homarr/validation";
 
 import { controlsInputSchema } from "../../../../integrations/src/pi-hole/pi-hole-types";
 import { createManyIntegrationMiddleware, createOneIntegrationMiddleware } from "../../middlewares/integration";
 import { createTRPCRouter, publicProcedure } from "../../trpc";
 
 export const dnsHoleRouter = createTRPCRouter({
   summary: publicProcedure
+    .input(z.object({ widgetKind: z.enum(["dnsHoleSummary", "dnsHoleControls"]) }))
     .unstable_concat(createManyIntegrationMiddleware("query", ...getIntegrationKindsByCategory("dnsHole")))
-    .query(async ({ ctx }) => {
+    .query(async ({ input: { widgetKind }, ctx }) => {
       const results = await Promise.all(
-        ctx.integrations.map(async (integration) => {
-          const cache = createCacheChannel<DnsHoleSummary>(`dns-hole-summary:${integration.id}`);
-          const { data } = await cache.consumeAsync(async () => {
-            const client = integrationCreator(integration);
-
-            return await client.getSummaryAsync().catch((err) => {
-              logger.error("dns-hole router - ", err);
-              throw new TRPCError({
-                code: "INTERNAL_SERVER_ERROR",
-                message: `Failed to fetch DNS Hole summary for ${integration.name} (${integration.id})`,
-              });
-            });
-          });
+        ctx.integrations.map(async ({ decryptedSecrets: _, ...integration }) => {
+          const channel = createItemAndIntegrationChannel<DnsHoleSummary>(widgetKind as WidgetKind, integration.id);
```

Comment:
Here and line 48, had to pass the widget kind from the widget calling it. would be nice to have it from context.
This can be done on the cron job side without a hitch, which is nice, but not here at first glance.
@Meierschlumpf

Reply:
Don't see a simple way right now